Mathlib/Order/Quantale.lean

```diff
@@ -0,0 +1,128 @@
+/-
+Copyright (c) 2024 Pieter Cuijpers. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pieter Cuijpers
+-/
+import Mathlib.Order.CompleteLattice
+import Mathlib.Algebra.Group.Defs
+
+/-!
+# Theory of quantales
+
+Quantales are the non-commutative generalization of locales/frames and as such are linked
+to point-free topology and order theory. Applications are found throughout logic,
+quantum mechanics, and computer science. Intuitively, as described by [vickers1989],
+open sets of a topology form a frame, and can be considered as modeling what is `observable`
+in a system. Quantales come into play when making an observation may change the system itself.
+Traditionally, one would write `x ⊓ y` to describe making observations `x` and `y` at the same
+time, but when the order of observation becomes important due to changes in the system caused
+by the observation, it makes more sense to write `x * y` or `x + y` and consider `*` or `+` as
+a possibly non-commutative monoid.
```

Comment:
Unless you use `to_additive`, you're only defining multiplicative quantales below. Maybe you should do both? (Note: you'll need to provide the `class` manually first, and then mark the multiplicative one with `to_additive`. See for example, how `Monoid` and `AddMonoid` are declared.)

Reply:
Thanks, that is indeed what I want ultimately. But since this is my first PR, I figured I should keep it simple and get used to the process first.

Also, when I was looking at AddMonoid, it was not completely clear to me yet how to_additive works. But that's a next step to figure out for me.

Reply:
`to_additive` should mostly "Just Work", except that for `class`es and `structure`s you have to declare them manually separately first (additive one first so that when you add the attribute to the multiplicative one it sees the existing additive declaration). Mostly, it works by finding the right name of the additive lemma and then inserting that everywhere. There is a dictionary for this, which isn't too hard to guess (`one` → `zero`, `mul` → `add`, `pow` → `smul`, etc.). There are edge cases where you need to do something manually, but if you have any trouble, just ask on Zulip.

Reply:
Ah, apparently I needed to update ToAdditive.Frontend as well, adding `Quantale` as a recognized thing to which to_additive applies. After that, things went smoothly indeed.